crates/bpe/README.md

```diff
@@ -183,40 +183,82 @@ On average it is about ~4 faster, since the short-cuts usually pay off.
 
-Results for counting o200k tokens for random 10000 byte slices. The setup time of the interval encoder is comparable to backtracking. After setup counting of slices of the original data are approximately constant time.
+- The second measures incremental encoding runtime, where the text is built up byte-by-byte.
+  This mode is not available in tiktoken, which only supports counting/encoding a complete text.
 
-![counting runtime comparison](./benches/result/counting-o200k.svg)
+- The third measures interval counting runtime, where the token count for slices of an original text are determined.
+  After the initial tokenization of the text, token counting for slices is typically constant time.
+  This mode is not available in tiktoken, which only supports counting/encoding a complete text.
+
+All benchmarks were run on a MacBook Pro M1.
+
+### Encoding
+
+Encoding is computing the tokens for a given text.
+This benchmark uses several encoders:
 
-### Encoding results
+- The backtracking encoder uses a backtracking algorithm based on a string matching automaton.
+- The heap encoder uses a priority heap to implement the traditional BPE algorithm.
+- The table encoder uses a dynamic programming algorithm.
 
-Results for encoding o200k tokens for random 1000 bytes. The backtracking encoder consistently outperforms tiktoken by a constant factor.
+Two additional encoders are included that are faster but do not always give exact results:
+
+- The greedy encoder uses a left-to-right greedy algorithm.
+- The minimal encoder computes an encoding with the minimal number of tokens.
+
+The benchmark measured the runtime of encoding of slices of lengths 10, 100, 1000, and 1000 from a random 20000 token original text using the o200k token set.
+(All encodings were computed from scratch for each slice.)
+
+The graph below shows encoding runtime vs slice length.
+All encoders show similar runtime increases with increasing slice length.
+The backtracking encoder, the fastest encoder that still returns correct results, shows a performance gain of approximately 3.5x compared to tiktoken.
+The fully dynamic programming solution and the heap implementation are still quite competitive to TikToken (especially for smaller inputs).
+If the requirement of correct BPE output can be relaxed, then the Greedy approach or the minimal encoding approach are the clear winners.
 
 ![encoding runtime comparison](./benches/result/encoding-o200k.svg)
 
-### Incremental encoding results
+### Incremental encoding
+
+Incremental encoding tokenizes a text to which bytes are appended.
+This benchmark uses two encoders:
 
-Results for incrementally encoding o200k tokens by appending 10000 random bytes. The appending encoder is slower by a constant factor but overall has similar performance curve as the backtracking encoder encoding all data at once.
+- The backtracking encoder, which retokenizes the text froms cratch every time it changes.
```

Comment:
mmm. this statement is not correct :)
Whenever it detects that it made a decision which is no longer compatible with new bytes, then it backtracks.
But it backtracks with memorisation which means it tries to compute only those states from the dynamic programming encoder which are actually necessary to find the correct solution...